cashier-google-play-billing/src/main/java/com/getkeepsafe/cashier/billing/InventoryQuery.java

```diff
@@ -0,0 +1,235 @@
+import com.android.billingclient.api.SkuDetails;
+import com.android.billingclient.api.SkuDetailsResponseListener;
+import com.getkeepsafe.cashier.Inventory;
+import com.getkeepsafe.cashier.InventoryListener;
+import com.getkeepsafe.cashier.Product;
+import com.getkeepsafe.cashier.Purchase;
+import com.getkeepsafe.cashier.Vendor;
+import com.getkeepsafe.cashier.VendorConstants;
+
+import org.json.JSONException;
+
+import java.util.ArrayList;
+import java.util.Collection;
+import java.util.Collections;
+import java.util.HashMap;
+import java.util.HashSet;
+import java.util.List;
+import java.util.Map;
+import java.util.Set;
+
+
+/**
+ * Inventory query helper class. Performs api calls to get requested products info and purchases.
+ * Since cashier purchase contains full product info, but Google Billing only returns order id and
+ * receipt, getSkuDetails call must be performed for all purchased skus.
+ */
+class InventoryQuery {
+
+    private InventoryListener listener;
+
+    private AbstractGooglePlayBillingApi api;
+
+    /**
+     * Inapp product details returned from async getSkuDetails call
+     * Non null value indicates that sku details query finished
+     */
+    private List<SkuDetails> inappSkuDetails = null;
+
+    /**
+     * Subscription product details returned from async getSkuDetails call
+     * Non null value indicates that sku details query finished
+     * */
+    private List<SkuDetails> subsSkuDetails = null;
+
+    /**
+     * List of purchases of both inapp and subscription types
+     * Non null value indicates that purchases query finished
+     */
+    private List<com.android.billingclient.api.Purchase> purchases = null;
+
+    private Collection<String> inappSkus;
+
+    private Collection<String> subSkus;
+
+    private int inappResponseCode = 0;
+
+    private int subsResponseCode = 0;
+
+    private Handler mainHandler;
+
+    /**
+     * Query inventory.
+     * @param api Google Play Billing API instance.
+     * @param listener Listener to deliver success / error.
+     * @param inappSkus List of product skus of item type to query. May be null.
+     * @param subSkus List of product skus of subscription type to query. May be null.
+     */
+    static void execute(@NonNull AbstractGooglePlayBillingApi api, @NonNull InventoryListener listener, @Nullable Collection<String> inappSkus, @Nullable Collection<String> subSkus) {
+        new InventoryQuery(api, listener, inappSkus, subSkus).execute();
+    }
+
+    private InventoryQuery(@NonNull AbstractGooglePlayBillingApi api, @NonNull InventoryListener listener, @Nullable Collection<String> inappSkus, @Nullable Collection<String> subSkus) {
+        this.api = api;
+        this.listener = listener;
+        this.inappSkus = inappSkus;
+        this.subSkus = subSkus;
+        mainHandler = new Handler(Looper.getMainLooper());
+    }
+
+    private void execute() {
+
+        // Execute on new thread to avoid blocking UI thread
+        new Thread() {
```

Comment:
I think it would be better to use a `Executor` in this case since allocating threads are expensive. You might want to consider `Executors.newCachedThreadPool()` or `Executors.newFixedThreadPool()`.